source/extensions/filters/http/tap/tap_config_impl.cc

```diff
@@ -0,0 +1,120 @@
+           proto_match_config.http_match_config().request_match_config().headers()) {
+        match_config.request_config_.value().headers_to_match_.emplace_back(header_match);
+      }
+    }
+
+    if (proto_match_config.http_match_config().has_response_match_config()) {
+      match_config.response_config_ = ResponseMatchConfig();
+      for (const auto& header_match :
+           proto_match_config.http_match_config().response_match_config().headers()) {
+        match_config.response_config_.value().headers_to_match_.emplace_back(header_match);
+      }
+    }
+  }
+}
+
+void HttpTapConfigImpl::matchesRequestHeaders(const Http::HeaderMap& headers,
+                                              std::vector<MatchStatus>& statuses) {
+  ASSERT(match_configs_.size() == statuses.size());
+  for (uint64_t i = 0; i < match_configs_.size(); i++) {
+    statuses[i].request_matched_ = true;
+    if (match_configs_[i].request_config_.has_value()) {
+      statuses[i].request_matched_ = Http::HeaderUtility::matchHeaders(
+          headers, match_configs_[i].request_config_.value().headers_to_match_);
+    }
+  }
+}
+
+void HttpTapConfigImpl::matchesResponseHeaders(const Http::HeaderMap& headers,
+                                               std::vector<MatchStatus>& statuses) {
+  ASSERT(match_configs_.size() == statuses.size());
+  for (uint64_t i = 0; i < match_configs_.size(); i++) {
+    statuses[i].response_matched_ = true;
+    if (match_configs_[i].response_config_.has_value()) {
+      statuses[i].response_matched_ = Http::HeaderUtility::matchHeaders(
+          headers, match_configs_[i].response_config_.value().headers_to_match_);
+    }
+  }
+}
+
+HttpPerRequestTapperPtr HttpTapConfigImpl::newPerRequestTapper() {
+  return std::make_unique<HttpPerRequestTapperImpl>(shared_from_this());
+}
+
+void HttpPerRequestTapperImpl::onRequestHeaders(const Http::HeaderMap& headers) {
+  config_->matchesRequestHeaders(headers, statuses_);
+}
+
+void HttpPerRequestTapperImpl::onResponseHeaders(const Http::HeaderMap& headers) {
+  config_->matchesResponseHeaders(headers, statuses_);
+}
+
+namespace {
+Http::HeaderMap::Iterate fillHeaderList(const Http::HeaderEntry& header, void* context) {
+  Protobuf::RepeatedPtrField<envoy::api::v2::core::HeaderValue>& header_list =
+      *reinterpret_cast<Protobuf::RepeatedPtrField<envoy::api::v2::core::HeaderValue>*>(context);
+  auto& new_header = *header_list.Add();
+  new_header.set_key(header.key().c_str());
+  new_header.set_value(header.value().c_str());
+  return Http::HeaderMap::Iterate::Continue;
+}
+} // namespace
+
+bool HttpPerRequestTapperImpl::onLog(const Http::HeaderMap* request_headers,
+                                     const Http::HeaderMap* response_headers) {
+  absl::optional<uint64_t> match_index;
+  for (uint64_t i = 0; i < statuses_.size(); i++) {
+    const auto& status = statuses_[i];
+    if (status.request_matched_ && status.response_matched_) {
+      match_index = i;
+      break;
```

Comment:
btw, is it guaranteed only one will match, or we only send the request to first config?

Reply:
This is just checking for match specs that matched. It is possible that more than one matched if there are multiple OR conditions. Yes, in this case right now we just send the ID of the first one that matched. I suppose we could send an array of match IDs? Thoughts?

Reply:
I think I'm going to get rid of this and go back to a single match condition, with the possibility of a complex set of AND/OR/NOT per discussion with @htuch so I'm going to resolve this.